libraries/core_libs/consensus/src/transaction_manager/transaction_manager.cpp

```diff
@@ -140,26 +140,40 @@ std::shared_ptr<Transaction> TransactionManager::getTransaction(trx_hash_t const
   return db_->getTransaction(hash);
 }
 
-void TransactionManager::removeTransactionsFromPool(SharedTransactions const &trxs) {
+void TransactionManager::saveTransactionsFromDagBlock(SharedTransactions const &trxs) {
   // This lock synchronizes inserting and removing transactions from transactions memory pool with database insertion.
   // Unique lock here makes sure that transactions we are removing are not reinserted in transactions_pool_
   std::unique_lock transactions_lock(transactions_mutex_);
+  auto write_batch = db_->createWriteBatch();
   for (auto const &trx : trxs) {
-    nonfinalized_transactions_in_dag_.emplace(trx->getHash());
     if (transactions_pool_.erase(trx->getHash())) {
       LOG(log_dg_) << "Transaction " << trx->getHash() << " removed from trx pool";
       // Transactions are counted when included in DAG
       trx_count_++;
+      transaction_accepted_.emit(trx->getHash());
     }
-    transaction_accepted_.emit(trx->getHash());
+    // We only save transaction if it has not already been saved
+    if (!db_->transactionInDb(trx->getHash())) {
+      db_->addTransactionToBatch(*trx, write_batch);
+      nonfinalized_transactions_in_dag_.emplace(trx->getHash());
+    }
+    db_->commitWriteBatch(write_batch);
   }
   db_->saveStatusField(StatusDbField::TrxCount, trx_count_);
 }
 
 void TransactionManager::recoverNonfinalizedTransactions() {
   // On restart populate nonfinalized_transactions_in_dag_ structure from db
+  auto write_batch = db_->createWriteBatch();
   for (auto const &trx : db_->getNonfinalizedTransactions()) {
-    nonfinalized_transactions_in_dag_.emplace(trx->getHash());
+    if (db_->transactionFinalized(trx->getHash())) {
+      // TODO: This section where transactions are deleted is only to recover from a bug where non-finalized
+      // transactions were not properly deleted from the DB, once there are no nodes with such corrupted data, this
+      // line can be removed oe replaced with an assert
+      db_->removeTransactionToBatch(trx->getHash(), write_batch);
+    } else {
+      nonfinalized_transactions_in_dag_.emplace(trx->getHash());
```

Comment:
there is no lock for nonfinalized_transactions_in_dag_, are we sure there is no possible concurrent access ?

Reply:
I fixed all your comments. This one I don't think there is a need to protect it since it is called from constructor but I added the lock anyway since we do protect it on all other places.